fuse_core/src/callback_wrapper.cpp

```diff
@@ -0,0 +1,145 @@
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <fuse_core/callback_wrapper.h>
+
+namespace fuse_core
+{
+
+
+  CallbackAdapter::CallbackAdapter(std::shared_ptr<rclcpp::Context> context_ptr){
+
+    rcl_guard_condition_options_t guard_condition_options =
+        rcl_guard_condition_get_default_options();
+
+    // Guard condition is used by the wait set to handle execute-or-not logic
+    gc_ = rcl_get_zero_initialized_guard_condition();
+    if (RCL_RET_OK != rcl_guard_condition_init(
+      &gc_, context_ptr->get_rcl_context().get(), guard_condition_options)
+    ) {
+      RCLCPP_WARN(rclcpp::get_logger("fuse"),
+                  "Could not init guard condition for callback waitable.");
+    }
+  }
+
+  /**
+   * @brief tell the CallbackGroup how many guard conditions are ready in this waitable
+   */
+  size_t CallbackAdapter::get_number_of_ready_guard_conditions() { return 1;}
+
+
+  /**
+   * @brief tell the CallbackGroup that this waitable is ready to execute anything
+   */
+  bool CallbackAdapter::is_ready(rcl_wait_set_t * wait_set) {
+    (void) wait_set;
+    // std::lock_guard<std::recursive_mutex> lock(queue_mutex_);
+    // a thread glitch isn't a disaster here and a mutex wouldn't stop it anyway
+    return !callback_queue_.empty();
+  }
+
+
+  /**
+   * @brief add_to_wait_set is called by rclcpp during NodeWaitables::add_waitable() and CallbackGroup::add_waitable()
+    waitable_ptr = std::make_shared<CallbackAdapter>();
+    node->get_node_waitables_interface()->add_waitable(waitable_ptr, (rclcpp::CallbackGroup::SharedPtr) nullptr);
+   */
+  void CallbackAdapter::add_to_wait_set(rcl_wait_set_t * wait_set)
+  {
+    std::lock_guard<std::recursive_mutex> lock(reentrant_mutex_);
+    RCLCPP_WARN(rclcpp::get_logger("fuse"),
+    "Could not add callback waitable to wait set.");
+    if (RCL_RET_OK != rcl_wait_set_add_guard_condition(wait_set, &gc_, NULL)) {
+    }
+  }
+
+  /**
+   * @brief unused part of the rclcpp::waitables interface
+   *
+   */
+  std::shared_ptr< void > CallbackAdapter::take_data(){
+    return nullptr;
```

Comment:
The idea behind `take_data` is the executor asks `rcl_wait` if there's any data to execute, finds out that the waitable is ready, takes the data from the lower level, and then passes it to the thing that's going to execute it.

Imagine there are multiple callbacks in the queue owned by `CallbackAdapter`, and a multithreaded executor is used. With the current implementation the executor discovers one callback is ready, queues it to execute, but doesn't learn that there's another callback until the first callback starts executing. This increases the time it takes for all callbacks after the first one to start, and the delay increases for each additional callback.

The fix is to make this the method where `callback_queue` is popped from. It should return the callback as a `std::shared_ptr<void>` that was created using [the `aliasing constructor`](https://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr). This pointer will later be fed as the parameter to `execute()`. This allows the executor to start executing all callbacks rapidly without waiting for the first one to start executing.

Reply:
I think I was reading from the executor pattern in rclc, it's good to hear the additional context

Reply:
https://github.com/locusrobotics/fuse/pull/284/commits/fb5254343a1098ead1608a3d65580615e611f1d6